smart_open/s3.py

```diff
@@ -545,38 +564,47 @@ def __init__(
         version_id=None,
         buffer_size=DEFAULT_BUFFER_SIZE,
         line_terminator=constants.BINARY_NEWLINE,
-        defer_seek=False,
+        stream_range=DEFAULT_STREAM_RANGE,
         client=None,
         client_kwargs=None,
     ):
+        assert isinstance(stream_range, int), \
+            'stream_range should be an integer number of bytes that restricts the maximum size the S3 server ' \
+            'needs to prepare for a single HTTP request. Got type: ' + type(stream_range)
+
         self._version_id = version_id
         self._buffer_size = buffer_size
 
-        _initialize_boto3(self, client, client_kwargs, bucket, key)
+        self._client, self._bucket, self._key = _initialize_boto3(client, client_kwargs, bucket, key)
+        self._version_id = version_id
 
-        self._raw_reader = _SeekableRawReader(
-            self._client,
-            bucket,
-            key,
-            self._version_id,
-        )
-        self._current_pos = 0
-        self._buffer = smart_open.bytebuffer.ByteBuffer(buffer_size)
+        self._content_length = None
         self._eof = False
+        self._position = 0
+        self._body = None
+        self._buffer = smart_open.bytebuffer.ByteBuffer(buffer_size)
         self._line_terminator = line_terminator
 
+        # the max_stream_size setting limits how much data will be read in a single HTTP request, this is an
+        # important protection for the S3 server, ensuring the S3 server doesn't get an open-ended byte-range request
+        # which can cause it to internally queue up a massive file when only a small bit of it may ultimately be
+        # read by the user. The variable _stream_range_[from|to] tracks the range of bytes that can be read
+        # from the current request body (e.g. from the same HTTP request). Note that the first read call
+        # will always set the byte range header to exactly the read size, an optimization for uses cases in which a
+        # single small read is performed against a large file (example: random sampling small data samples from
+        # large files in machine learning contexts).
+        self._stream_range = stream_range
+        self._stream_range_from = None  # a None value signifies the first call to `read` where this will be set
```

Comment:
I think the current name is confusing, it's too similar to self._stream_range.

```suggestion
        self._stream_position_from = None  # a None value signifies the first call to `read` where this will be set
```